File: src/Filter.java
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.util.*;
import org.json.simple.JSONObject;
import org.json.simple.JSONArray;
import org.json.simple.parser.JSONParser;
import org.json.simple.parser.ParseException;
import java.io.IOException; 


public class Filter {
    private int price = -1;
    private int area = -1;
    private int floor = -1;
    private int town = -1;
    private int model = -1;
    private long leaseStartDate = 0;
    public final int MIN_PRICE = 5000;
    public final int MAX_PRICE = 1418000;
    public final int MIN_FLOOR = 1;
    public final int MAX_FLOOR = 51;
    private JSONObject townData = new JSONObject();
    private JSONObject flatTypeData = new JSONObject();
    
    public Filter() {
        readJson();
    }

    public Filter(int price, int area, int floor, String town, String model, long leaseStartDate){
        readJson();
        this.price = price;
        this.area = area;
        this.floor = floor;
        this.town = convertName(1, town);
        this.model = convertName(2, model);
        this.leaseStartDate = leaseStartDate;
    }

    private void readJson()
    {
        FileReader dataFile;
        try {
            dataFile = new FileReader("resources/data/conversion.json");
            JSONParser jsonParser = new JSONParser();
            JSONObject ds;
            try {
                ds = (JSONObject) jsonParser.parse(dataFile);
                JSONArray tArray = new JSONArray();
                tArray = (JSONArray) (ds.get("town"));
                this.townData = (JSONObject) tArray.get(0);
                tArray.clear();
                tArray = (JSONArray) (ds.get("flatType"));
                this.flatTypeData = (JSONObject) tArray.get(0);
            } catch (IOException | ParseException e) {
                //e.printStackTrace();
                System.out.println("Parse Error");
            }
        } catch (FileNotFoundException e) {
            //e.printStackTrace();
            System.out.println("File not found");
        }
    }

    public void testFilter()
    {
        System.out.println("Town number: " + this.town);
        System.out.println("Price: " + this.price);
        System.out.println("Area: " + this.area);
        System.out.println("Floor: " + this.floor);
        System.out.println("Model: " + this.model);
        System.out.println("Lease Start Date: " + this.leaseStartDate);
    }

    private int convertName(int c, String data) {
        switch (c) {
            case 1:
                return Integer.parseInt(this.townData.get(data).toString());
            case 2:
                return Integer.parseInt(this.flatTypeData.get(data).toString());
            default:
                return -1;
        }
    }

    public int[] applyFilter() {
        int[] finalFilter = { price, area, floor, town, model, model, (int) leaseStartDate };
        return finalFilter;
    }
}

File: firebase/public/Filter.js
const fs = require('fs');
const express = require('express');
const cors = require('cors'); 
const bodyParser = require('body-parser');
const formula = require('./formula')
let conversion;
let townData;
let dateTime;
const app = express();
const router = express.Router();
app.use(express.json());
app.use(cors({ origin: 'http://localhost:3000', }));
var allowCrossDomain = function(req,res,next) {
    res.header('Access-Control-Allow-Origin', '*');
    res.header('Access-Control-Allow-Methods', 'GET, PUT, POST, DELETE');
    res.header('Access-Control-Allow-Headers', 'Content-Type');
    next();  
  }
  app.use(allowCrossDomain);

//Read conversion.json
fs.readFile("./resources/data/conversion.json", "utf8", (err, jsonString) => {
    if (err) {
        console.log("File read failed:", err);
        return;
    }
    conversion = JSON.parse(jsonString);
    townData = conversion.town;
    flatData = conversion.flatType;
    roomNoData = conversion.roomNumber;

    //TestProgram
    //test(); 
    //console.log(formula(Filter()));
});

let area = -1;
let floor = -1;
let town = -1;
let model = -1;
let roomNo = -1;
let leaseStartDate = 0;
const MIN_PRICE = 5000;
const MAX_PRICE = 1418000;
const MIN_FLOOR = 1;
const MAX_FLOOR = 51;
const averageFloor = [2, 3, 5, 8, 11, 13, 14, 17, 18, 20, 23, 26, 28, 29, 32, 33, 35, 38, 41, 44, 47, 50];

//set Filter
const SetFilter = (data) => {
    area = parseInt(data['floorArea']);
    floor = parseInt(data['floor']);
    for(let x = 0; x < averageFloor.length; x++)
    {
        if(averageFloor[x] == floor)
        {
            floor = averageFloor[x];
            break;
        }
        if(averageFloor[x] > floor)
        {
            if((averageFloor[x] - floor) >= (floor - averageFloor[x-1]) || x == 0)
            {
                floor = averageFloor[x];
                break;
            }
            else {
                floor = averageFloor[x - 1];
                break;
            }

        }
    }
    town = townData[0][data['townName']];
    model = flatData[0][data['flatModel']];
    roomNo = roomNoData[0][data['noOfRooms']];
    dateTime = new Date(data['leaseStartDate']);
    //leaseStartDate = Math.floor(dateTime.getTime()/1000); //UNIX TIME: DO NOT USE
    leaseStartDate = parseInt(data['leaseStartDate']);
}

//node js testing purposes
const test = () => {
    area = 100;
    floor = 6;
    town = townData[0]['CHOA CHU KANG'];
    model = flatData[0]['New Generation'];
    roomNo = roomNoData[0]['5 ROOM'];
    leaseStartDate = 1980;
}

//get Filter
const Filter = () => {
    let filterObj = { area, floor, town, model, roomNo, leaseStartDate };
    return filterObj;
}

app.use(bodyParser.urlencoded({ extended: false }));
app.use(bodyParser.text());

app.post("/filterReq", (req, res) => { 
const rawData = req.body;
SetFilter(rawData);
console.log('Data collected ' + Filter()['area']);
res.send({data: formula(Filter())});
console.log('Data sent ' + JSON.stringify(formula(Filter())));
res.end();
});

/*app.get("/test", (req, res) => {
    const rawData = req.body;
    //res.send({price: '1'});
    res.status(200).json({price: '1'});
});*/

app.listen(5001, () => {
    console.log(`Server is running on port 3001.`);
});
